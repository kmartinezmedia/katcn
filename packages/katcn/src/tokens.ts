import type {
  AvatarSizeConfig,
  BorderRadiusConfig,
  BorderWidthConfig,
  FontFamilyGlobalConfig,
  FontWeightDescriptive,
  FontWeightNumeric,
  HuesChromaConfig,
  HuesConfig,
  HuesLightnessConfig,
  IconSizeConfig,
  PaletteConfig,
  SpacingConfig,
  TypographyConfig,
  UniversalTokensConfig,
  ZIndexConfig,
} from './types';

const hues: HuesConfig = {
  magenta: 331.6,
  pink: 4.6,
  rose: 7.4,
  red: 27.4,
  sunset: 33.9,
  orange: 42.7,
  nude: 42.5,
  brown: 68.8,
  yellow: 82.9,
  citron: 109.8,
  lime: 133.2,
  green: 154.4,
  mint: 164.9,
  teal: 194.8,
  cyan: 224.1,
  blue: 261,
  indigo: 273.3,
  purple: 289,
  gray: 248.1,
  carbon: 0,
};

const huesLightness: HuesLightnessConfig = {
  '0': '99%',
  '1': '95%',
  '2': '88%',
  '3': '80%',
  '4': '74%',
  '5': '68%',
  '6': '63%',
  '7': '58%',
  '8': '53%',
  '9': '49%',
  '10': '42%',
  '11': '35%',
  '12': '27%',
  '13': '20%',
  '14': '14%',
  '15': '11%',
};

const huesChroma: HuesChromaConfig = {
  '0': 0.03,
  '1': 0.06,
  '2': 0.12,
  '3': 0.14,
  '4': 0.16,
  '5': 0.19,
  '6': 0.2,
  '7': 0.21,
  '8': 0.2,
  '9': 0.19,
  '10': 0.17,
  '11': 0.15,
  '12': 0.12,
  '13': 0.09,
  '14': 0.07,
  '15': 0.05,
};

const palette: PaletteConfig = {
  core: {
    brand: { hue: 'purple', step: 9 },
    accent: { hue: 'blue', step: 9 },
    alert: { hue: 'red', step: 9 },
    positive: { hue: 'green', step: 9 },
    warning: { hue: 'orange', step: 9 },
  },
  background: {
    primary: { hue: 'gray', step: 0 },
    secondary: { hue: 'gray', step: 2 },
  },
  foreground: {
    primary: { hue: 'gray', step: 13 },
    secondary: { hue: 'gray', step: 14 },
    tertiary: { hue: 'gray', step: 9 },
    'on-color': { hue: 'gray', step: 0 },
  },
  line: {
    primary: { hue: 'gray', step: 14 },
    secondary: { hue: 'gray', step: 9 },
  },
};

const sansFallbacks = ['Helvetica Neue', 'Helvetica', 'Arial', 'sans-serif'];
const serifFallbacks = ['Georgia', 'Times', 'Times New Roman', 'serif'];

const fontFamily: FontFamilyGlobalConfig = {
  icons: { fallbacks: ['icons', ...sansFallbacks], name: 'icons' },
  sans: { fallbacks: sansFallbacks, name: 'sans' },
  'sans-condensed': { fallbacks: sansFallbacks, name: 'sans-condensed' },
  'serif-text': { fallbacks: serifFallbacks, name: 'serif-text' },
  'serif-display': { fallbacks: serifFallbacks, name: 'serif-display' },
  mono: {
    fallbacks: ['Menlo', 'Monaco', 'Courier New', 'monospace'],
    name: 'mono',
  },
};

const typography: TypographyConfig = {
  display1: {
    fontFamily: 'sans',
    fontSize: '48px',
    lineHeight: '44px',
    fontWeight: 'black',
    textTransform: 'uppercase',
  },
  display2: {
    fontFamily: 'sans',
    fontSize: '40px',
    lineHeight: '44px',
    fontWeight: 'extrabold',
  },
  display3: {
    fontFamily: 'sans',
    fontSize: '36px',
    lineHeight: '44px',
    fontWeight: 'extrabold',
  },
  title1: {
    fontFamily: 'sans',
    fontSize: '32px',
    lineHeight: '40px',
    fontWeight: 'bold',
  },
  title2: {
    fontFamily: 'sans',
    fontSize: '28px',
    lineHeight: '36px',
    fontWeight: 'bold',
  },
  title3: {
    fontFamily: 'sans',
    fontSize: '24px',
    lineHeight: '32px',
    fontWeight: 'bold',
  },
  title4: {
    fontFamily: 'sans',
    fontSize: '20px',
    lineHeight: '24px',
    fontWeight: 'bold',
  },
  headline1: {
    fontFamily: 'sans',
    fontSize: '16px',
    lineHeight: '20px',
    fontWeight: 'semibold',
  },
  body1: {
    fontFamily: 'sans',
    fontSize: '16px',
    lineHeight: '20px',
  },
  label1: {
    fontFamily: 'sans',
    fontSize: '14px',
    lineHeight: '20px',
    fontWeight: 'semibold',
  },
  label2: {
    fontFamily: 'sans',
    fontSize: '14px',
    lineHeight: '20px',
  },
  caption1: {
    fontFamily: 'sans',
    fontSize: '12px',
    lineHeight: '16px',
    fontWeight: 'semibold',
  },
  caption2: {
    fontFamily: 'sans',
    fontSize: '12px',
    lineHeight: '16px',
  },
  legal1: {
    fontFamily: 'sans',
    fontSize: '11px',
    lineHeight: '16px',
    fontWeight: 'semibold',
  },
};

const borderRadius: BorderRadiusConfig = {
  none: '0px',
  xs: '2px',
  sm: '4px',
  md: '8px',
  lg: '16px',
  xl: '32px',
  full: '9999px',
};

const borderWidth: BorderWidthConfig = {
  none: '0px',
  thin: '1px',
  medium: '2px',
  thick: '4px',
};

const spacing: SpacingConfig = {
  px: '1px',
  '0': '0px',
  '0.5': '0.125rem',
  '1': '0.25rem',
  '1.5': '0.375rem',
  '2': '0.5rem',
  '2.5': '0.625rem',
  '3': '0.75rem',
  '3.5': '0.875rem',
  '4': '1rem',
  '5': '1.25rem',
  '6': '1.5rem',
  '7': '1.75rem',
  '8': '2rem',
  '9': '2.25rem',
  '10': '2.5rem',
  '11': '2.75rem',
  '12': '3rem',
  '14': '3.5rem',
  '16': '4rem',
  '20': '5rem',
  '24': '6rem',
  '28': '7rem',
  '32': '8rem',
  '36': '9rem',
  '40': '10rem',
  '44': '11rem',
  '48': '12rem',
  '52': '13rem',
  '56': '14rem',
  '60': '15rem',
  '64': '16rem',
  '72': '18rem',
  '80': '20rem',
  '96': '24rem',
};

const avatarSizes: AvatarSizeConfig = {
  sm: '16px',
  md: '24px',
  lg: '32px',
};

const iconSizes: IconSizeConfig = {
  sm: '16px',
  md: '24px',
  lg: '32px',
};

const zIndex: ZIndexConfig = {
  auto: 'auto',
  '0': '0',
  '10': '10',
  '20': '20',
  '30': '30',
  '40': '40',
  '50': '50',
};

export const defaultTokensConfig: UniversalTokensConfig = {
  palette,
  hues,
  huesChroma,
  huesLightness,
  borderRadius,
  borderWidth,
  spacing,
  avatarSizes,
  iconSizes,
  zIndex,
  fontFamily,
  typography,
};

export const alwaysPalette = {
  white: '#ffffff',
  black: '#000000',
  transparent: 'transparent',
};

export const fontWeightMap: Record<FontWeightDescriptive, FontWeightNumeric> = {
  thin: '100',
  extralight: '200',
  light: '300',
  regular: '400',
  medium: '500',
  semibold: '600',
  bold: '700',
  extrabold: '800',
  black: '900',
};
