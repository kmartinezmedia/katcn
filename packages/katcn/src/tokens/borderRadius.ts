import type { BorderRadiusConfig } from '../types';

const compactBorderRadius: BorderRadiusConfig = {
  none: 0,
  xs: 2,
  sm: 4,
  md: 8,
  lg: 12,
  xl: 24,
  full: 9999,
};

const normalBorderRadius: BorderRadiusConfig = {
  none: 0,
  xs: 2,
  sm: 4,
  md: 8,
  lg: 16,
  xl: 32,
  full: 9999,
};

export { compactBorderRadius, normalBorderRadius };
