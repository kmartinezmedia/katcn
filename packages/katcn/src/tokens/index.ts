export * from './borderRadius';
export * from './borderWidth';
export * from './defaultTokensConfig';
export * from './elevation';
export * from './layout';
export * from './motion';
export * from './palette';
export * from './sizes';
export * from './spacing';
export * from './spectrum';
export * from './typography';
