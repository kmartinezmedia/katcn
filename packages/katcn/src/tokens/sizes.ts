import type { AvatarSizeConfig, IconSizeConfig } from '../types';

const compactAvatarSizes: AvatarSizeConfig = {
  sm: 16,
  md: 24,
  lg: 32,
};

const normalAvatarSizes: AvatarSizeConfig = compactAvatarSizes;

const compactIconSizes: IconSizeConfig = {
  sm: 16,
  md: 24,
  lg: 32,
};

const normalIconSizes: IconSizeConfig = compactIconSizes;

export {
  compactAvatarSizes,
  compactIconSizes,
  normalAvatarSizes,
  normalIconSizes,
};
