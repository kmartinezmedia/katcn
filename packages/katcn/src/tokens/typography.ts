import type {
  FontFamilyConfig,
  FontFamilyGlobalConfig,
  FontSizeConfig,
  FontWeightConfig,
  FontWeightDescriptive,
  FontWeightNumeric,
  LineHeightConfig,
  TextTransformConfig,
} from '../types';

const sansFallbacks = ['Helvetica Neue', 'Helvetica', 'Arial', 'sans-serif'];
const serifFallbacks = ['Georgia', 'Times', 'Times New Roman', 'serif'];

export const fontFamilyGlobal: FontFamilyGlobalConfig = {
  icons: { fallbacks: ['icons', ...sansFallbacks], name: 'icons' },
  sans: { fallbacks: sansFallbacks, name: 'sans' },
  'sans-condensed': { fallbacks: sansFallbacks, name: 'sans-condensed' },
  'serif-text': { fallbacks: serifFallbacks, name: 'serif-text' },
  'serif-display': { fallbacks: serifFallbacks, name: 'serif-display' },
  mono: {
    fallbacks: ['Menlo', 'Monaco', 'Courier New', 'monospace'],
    name: 'mono',
  },
};

export const fontFamily: FontFamilyConfig = {
  display1: 'sans',
  display2: 'sans',
  display3: 'sans',
  title1: 'sans',
  title2: 'sans',
  title3: 'sans',
  title4: 'sans',
  headline1: 'sans',
  body1: 'sans',
  label1: 'sans',
  label2: 'sans',
  caption1: 'sans',
  caption2: 'sans',
  legal1: 'sans',
};

export const fontWeightMap: Record<FontWeightDescriptive, FontWeightNumeric> = {
  thin: '100',
  extralight: '200',
  light: '300',
  regular: '400',
  medium: '500',
  semibold: '600',
  bold: '700',
  extrabold: '800',
  black: '900',
};

export const fontWeight: FontWeightConfig = {
  display1: 'black',
  display2: 'extrabold',
  display3: 'extrabold',
  title1: 'bold',
  title2: 'bold',
  title3: 'bold',
  title4: 'bold',
  headline1: 'semibold',
  body1: 'regular',
  label1: 'semibold',
  label2: 'regular',
  caption1: 'semibold',
  caption2: 'regular',
  legal1: 'semibold',
};

export const textTransform: TextTransformConfig = {
  display1: 'uppercase',
  display2: 'none',
  display3: 'none',
  title1: 'none',
  title2: 'none',
  title3: 'none',
  title4: 'none',
  headline1: 'none',
  body1: 'none',
  label1: 'none',
  label2: 'none',
  caption1: 'none',
  caption2: 'none',
  legal1: 'none',
};

export const xSmall = {
  fontSize: {
    display1: 46,
    display2: 37,
    display3: 33,
    title1: 29,
    title2: 25,
    title3: 21,
    title4: 17,
    headline1: 13,
    body1: 13,
    label1: 12,
    label2: 12,
    caption1: 11,
    caption2: 11,
    legal1: 10,
  } satisfies FontSizeConfig,
  lineHeight: {
    display1: 44,
    display2: 44,
    display3: 44,
    title1: 36,
    title2: 32,
    title3: 28,
    title4: 24,
    headline1: 16,
    body1: 16,
    label1: 16,
    label2: 16,
    caption1: 16,
    caption2: 16,
    legal1: 12,
  } satisfies LineHeightConfig,
};

export const small = {
  fontSize: {
    display1: 46,
    display2: 38,
    display3: 34,
    title1: 30,
    title2: 26,
    title3: 22,
    title4: 18,
    headline1: 14,
    body1: 14,
    label1: 13,
    label2: 13,
    caption1: 12,
    caption2: 12,
    legal1: 11,
  } satisfies FontSizeConfig,
  lineHeight: {
    ...xSmall.lineHeight,
    title1: 40,
    headline1: 20,
    body1: 20,
    legal1: 16,
  } satisfies LineHeightConfig,
};

export const medium = {
  fontSize: {
    display1: 47,
    display2: 39,
    display3: 35,
    title1: 31,
    title2: 27,
    title3: 23,
    title4: 19,
    headline1: 15,
    body1: 15,
    label1: 13,
    label2: 13,
    caption1: 12,
    caption2: 12,
    legal1: 11,
  } satisfies FontSizeConfig,
  lineHeight: {
    ...small.lineHeight,
    title2: 36,
    title3: 32,
  } satisfies LineHeightConfig,
};

export const large = {
  fontSize: {
    ...medium.fontSize,
    display1: 48,
    display2: 40,
    display3: 36,
    title1: 32,
    title2: 28,
    title3: 24,
    title4: 20,
    headline1: 16,
    body1: 16,
    label1: 14,
    label2: 14,
  } satisfies FontSizeConfig,
  lineHeight: {
    ...medium.lineHeight,
    title2: 36,
    title3: 32,
    headline1: 20,
    body1: 20,
    label1: 20,
    label2: 20,
  } satisfies LineHeightConfig,
};

export const xLarge = {
  fontSize: {
    display1: 50,
    display2: 42,
    display3: 38,
    title1: 34,
    title2: 30,
    title3: 26,
    title4: 22,
    headline1: 18,
    body1: 16,
    label1: 16,
    label2: 16,
    caption1: 14,
    caption2: 14,
    legal1: 13,
  } satisfies FontSizeConfig,
  lineHeight: {
    ...large.lineHeight,
    title1: 44,
    headline1: 24,
    body1: 24,
    caption1: 20,
    caption2: 20,
  } satisfies LineHeightConfig,
};

export const xxLarge = {
  fontSize: {
    display1: 52,
    display2: 44,
    display3: 40,
    title1: 36,
    title2: 32,
    title3: 28,
    title4: 26,
    headline1: 20,
    body1: 20,
    label1: 18,
    label2: 18,
    caption1: 16,
    caption2: 16,
    legal1: 15,
  } satisfies FontSizeConfig,
  lineHeight: {
    ...xLarge.lineHeight,
    title2: 40,
    title3: 36,
    headline1: 28,
    body1: 28,
    label1: 24,
    label2: 24,
    legal1: 20,
  } satisfies LineHeightConfig,
};

export const xxxLarge = {
  fontSize: {
    display1: 54,
    display2: 46,
    display3: 42,
    title1: 38,
    title2: 34,
    title3: 30,
    title4: 28,
    headline1: 22,
    body1: 22,
    label1: 20,
    label2: 20,
    caption1: 18,
    caption2: 18,
    legal1: 17,
  } satisfies FontSizeConfig,
  lineHeight: {
    ...xxLarge.lineHeight,
    title1: 48,
    title2: 44,
    title4: 32,
    label1: 28,
    label2: 28,
    caption1: 24,
    caption2: 24,
  } satisfies LineHeightConfig,
};
