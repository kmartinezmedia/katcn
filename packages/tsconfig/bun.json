{
  "$schema": "https://json.schemastore.org/tsconfig",
  "display": "Bun",
  "extends": "./node.json",
  "compilerOptions": {
    "types": ["bun-types"]
  }
}
