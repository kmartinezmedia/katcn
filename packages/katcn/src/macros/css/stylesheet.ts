import { cssEscape, entries, flattenObj } from '../../helpers';
import { defaultTokensConfig } from '../../tokens';
import type { ColorMode, UniversalTokensConfig } from '../../types';
import { CSS_VAR_PREFIX } from './consts';
import { createTheme } from './createTheme';
import { createPreflight } from './createPreflight';
import { createUtilities } from './createUtilities';

import {
  type CustomAtRules,
  Features,
  type Visitor,
  composeVisitors,
  transform,
} from 'lightningcss';

const decoder = new TextDecoder();

type StyleMap = Record<string, string | Record<string, string>>;

export interface KatcnStyleSheetOpts {
  disablePreflight: boolean;
  config: UniversalTokensConfig;
}

function processLayer(layer: Set<string>) {
  return Array.from(layer).join('\n');
}

function convertArbitraryDefinition(
  input: string,
): { [key: string]: string } | null {
  // Regular expression to extract the key and value
  const regex = /(\w+)-\[(\d+px)\]/;
  const match = input.match(regex);

  if (match) {
    // match[1] contains the key (e.g., "height")
    // match[2] contains the value (e.g., "200px")
    const key = match[1];
    const value = match[2];

    // Constructing the object dynamically
    const result = { [key]: value };
    return result;
  }

  return null; // Return null if no match was found
}

export class KatcnStyleSheet {
  public config: UniversalTokensConfig = defaultTokensConfig;
  public classNamesToAdd = new Map<string, Set<string>>();
  public classNamesToKeep = new Map<string, Set<string>>();
  public varsToKeep = new Set<string>();

  public base = new Set<string>();
  public theme = new Set<string>();
  public utilities = new Set<string>();

  private addThemeVars = this.addVarsToLayer(this.theme);
  private addUtilClasses = this.addClassesToLayer(this.utilities);

  constructor(public opts: KatcnStyleSheetOpts) {
    if (opts.config) {
      this.config = opts.config;
    }

    if (!opts?.disablePreflight) {
      this.base.add(createPreflight());
    }

    this.addThemeVars(flattenObj(createTheme(this.config)));
    this.addUtilClasses(this.utilClasses);
  }

  get allClassNamesToKeep() {
    const allClassNames = new Set<string>();

    for (const classNames of this.classNamesToKeep.values()) {
      for (const className of classNames) {
        const splitClassName = className.trimStart().trimEnd().split(' ');
        for (const sClassName of splitClassName) {
          allClassNames.add(sClassName);
        }
      }
    }

    return allClassNames;
  }

  get allClassNamesToAdd() {
    const allClassNames = new Set<string>();

    for (const classNames of this.classNamesToAdd.values()) {
      for (const className of classNames) {
        const splitClassName = className.trimStart().trimEnd().split(' ');
        for (const sClassName of splitClassName) {
          if (sClassName.includes('-[')) {
            allClassNames.add(sClassName);
          }
        }
      }
    }

    return allClassNames;
  }

  get utilClasses() {
    const utilsObj = flattenObj(createUtilities());
    if (this.allClassNamesToAdd.size > 0) {
      for (const className of this.allClassNamesToAdd) {
        const value = convertArbitraryDefinition(className);
        if (value) {
          let valueString = '';
          for (const [key, val] of entries(value)) {
            valueString += `${key}: ${val};`;
          }
          utilsObj[className] = `{ ${valueString} }`;
        }
      }
    }
    return utilsObj;
  }

  addVarToLayer(layer: Set<string>) {
    return (property: string, value: string) => {
      layer.add(`--${CSS_VAR_PREFIX}-${property}: ${value};`);
    };
  }

  addVarsToLayer(layer: Set<string>) {
    return (stylemap: StyleMap) => {
      const flattenedObj = flattenObj(stylemap);
      for (const [property, value] of entries(flattenedObj)) {
        const escapedVar = cssEscape(property);
        layer.add(`--${CSS_VAR_PREFIX}-${escapedVar}: ${value};`);
      }
    };
  }

  addClassesToLayer(layer: Set<string>) {
    return (stylemap: StyleMap) => {
      const flattenedObj = flattenObj(stylemap);
      for (const [classname, definition] of entries(flattenedObj)) {
        const escapedClassname = cssEscape(classname);
        layer.add(`.${escapedClassname} ${definition}`);
      }
    };
  }

  get cssTemplate() {
    return `
      @layer base {
        ${processLayer(this.base)}
      }

      @layer theme {
        :where(html) {
          ${processLayer(this.theme)}
        }
      }

      @layer utilities {
        ${processLayer(this.utilities)}
      }
    `;
  }

  get ruleVisitor() {
    const allClassNamesToKeep = this.allClassNamesToKeep;
    const allClassNamesToAdd = this.allClassNamesToAdd;
    const varsToKeep = this.varsToKeep;
    return {
      Rule(rule) {
        if (rule.type === 'style') {
          // Filter selectors (i.e. classnames) and keep only the ones in the safelist
          for (const selector of rule.value.selectors) {
            for (const sel of selector) {
              if (sel.type === 'attribute') {
                return rule;
              }

              if (sel.type === 'class') {
                if (
                  allClassNamesToKeep?.has(sel.name) ||
                  allClassNamesToAdd?.has(sel.name)
                ) {
                  for (const decl of rule.value.declarations.declarations) {
                    if (decl.property === 'unparsed') {
                      for (const val of decl.value.value) {
                        if (val.type === 'var') {
                          const varName = val.value.name.ident;
                          varsToKeep.add(varName);
                        }
                      }
                    }
                  }
                  return rule;
                }
                // Discard any unused classNames
                return [];
              }
            }
          }
          // Keep all other rules like @base, @theme, @utilities
          return rule;
        }
      },
    } satisfies Visitor<CustomAtRules>;
  }

  declarationVisitor(varsToKeep: Set<string>) {
    return {
      Declaration(decl) {
        if (decl.property === 'custom') {
          const declarationName = decl.value.name;
          const declarationValue = decl.value.value;

          if (varsToKeep.has(declarationName)) {
            for (const val of declarationValue) {
              if (val.type === 'function' && val.value.name === 'oklch') {
                for (const varInOklch of val.value.arguments) {
                  if (varInOklch.type === 'var') {
                    const varName = varInOklch.value.name.ident;
                    varsToKeep.add(varName);
                  }
                }
              }
              if (val.type === 'var') {
                const parentVarName = val.value.name.ident;
                varsToKeep.add(parentVarName);
              }
            }
            return decl;
          }

          return [];
        }
      },
    } satisfies Visitor<CustomAtRules>;
  }

  get css() {
    /** Lightning css to purge final stylesheet */
    // use safelist to keep classes only the used classnames
    transform({
      filename: 'pass1.css',
      code: Buffer.from(this.cssTemplate),
      minify: true,
      visitor: this.ruleVisitor,
    });

    transform({
      filename: 'pass2.css',
      code: Buffer.from(this.cssTemplate),
      minify: true,
      visitor: this.declarationVisitor(this.varsToKeep),
    });

    transform({
      filename: 'pass3.css',
      code: Buffer.from(this.cssTemplate),
      minify: true,
      visitor: this.declarationVisitor(this.varsToKeep),
    });

    const transformedCss = transform({
      filename: 'style.css',
      code: Buffer.from(this.cssTemplate),
      minify: true,
      include: Features.OklabColors | Features.LogicalProperties,
      visitor: composeVisitors([
        this.ruleVisitor,
        this.declarationVisitor(this.varsToKeep),
      ]),
    });

    return decoder.decode(transformedCss.code);
  }
}
