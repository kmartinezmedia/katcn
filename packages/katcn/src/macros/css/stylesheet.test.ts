import { describe, expect, it } from 'bun:test';
import { KatcnStyleSheet } from './stylesheet';
import { defaultTokensConfig } from '../../tokens';

describe('KatcnStyleSheet', () => {
  it('should include preflight', () => {
    const stylesheet = new KatcnStyleSheet({
      config: defaultTokensConfig,
      disablePreflight: false,
    });
    expect(stylesheet.cssTemplate).toBeDefined();
  });

  it('should return full stylesheet in cssTemplate', () => {
    const stylesheet = new KatcnStyleSheet({
      config: defaultTokensConfig,
      disablePreflight: true,
    });
    expect(stylesheet.cssTemplate).toBeDefined();
  });

  it('should correctly include arbitrary values', () => {
    const stylesheet = new KatcnStyleSheet({
      config: defaultTokensConfig,
      disablePreflight: true,
    });
    const fileRegistry = new Set<string>();
    fileRegistry.add('backgroundColor-accent');
    fileRegistry.add('width-[100px]');
    stylesheet.safelist.set('some-file.tsx', fileRegistry);
    expect(stylesheet.css).toBeDefined();
    expect(stylesheet.css.includes('backgroundColor-accent')).toBeTrue();
    // TODO: figure out why can't use full escaped classname to check
    expect(stylesheet.css.includes('.width-')).toBeTrue();
  });

  /** TODO: Convert stylesheet to object to be able to validate output */
});
