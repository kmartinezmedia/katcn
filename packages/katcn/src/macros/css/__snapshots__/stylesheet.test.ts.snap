// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`KatcnStyleSheet should correctly purge css 1`] = `"@layer theme{:where(html){--katcn-hue-blue:261;--katcn-hue-lightness-9:49.0%;--katcn-hue-chroma-9:.19;--katcn-color-blue-9:oklch(var(--katcn-hue-lightness-9)var(--katcn-hue-chroma-9)var(--katcn-hue-blue));--katcn-palette-core-accent:var(--katcn-color-blue-9)}}@layer utilities{.backgroundColor-accent{background-color:var(--katcn-palette-core-accent)}.width-\[100px\]{width:100px}}"`;
